Make System.Transactions.Local trimmable on Windows

eerhardt opened the pull request:
Remove `IsTrimmable=false` from the project, so this assembly is still trimmed with `partial` trimming on Windows.

Add a "LibraryBuild" ILLink warning, so when the distributed transaction code is not trimmed, the app developer gets a warning that it is not guaranteed to work.

Fix #75031

The full, correct fix here would be to use COMWrappers. However, https://github.com/dotnet/runtime/issues/75031#issuecomment-1237882011 indicates that level of investment isn't justified at this time. So using a temporary approach for now.

eerhardt committed:
Make System.Transactions.Local trimmable on Windows

Add the necessary DynamicDependency attributes to System.Transaction.Local in order for it to work with trimming on Windows. Also fix the UnconditionalSuppressMessage attribute to have a valid justification for suppressing, now that we are preserving the necessary interfaces for COM Interop.

Fix #75031

jkotas commented on src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:
```diff
@@ -47,7 +47,28 @@ internal DtcProxyShimFactory(EventWaitHandle notificationEventHandle)
         [MarshalAs(UnmanagedType.Interface)] out ITransactionDispenser ppvObject);
 
-    [UnconditionalSuppressMessage("Trimming", "IL2050", Justification = "Leave me alone")]
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2050:COMMarshalling",
```
I do not think this is sufficient to guarantee that Com works.

The built-in Com marshalling also depends on a bunch of internal runtime helper types that may get stripped.

jkotas replied:
The rest of the change is fine. I think this suppression should be deleted and we should leave the warning on.

MichalStrehovsky replied:
We should mark this `RequiresUnreferencedCode` with a message saying the current implementation is not compatible with trimming.

eerhardt replied:
I've updated the code to leave the warning in the library. If an app uses this code while trimming they will get a warning.

eerhardt committed:
Leave the ILLink warning in the product code so app developers get a warning that Distributed transactions don't work with trimming.

eerhardt commented on src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:
```diff
@@ -47,7 +47,30 @@ internal DtcProxyShimFactory(EventWaitHandle notificationEventHandle)
         [MarshalAs(UnmanagedType.Interface)] out ITransactionDispenser ppvObject);
 
-    [UnconditionalSuppressMessage("Trimming", "IL2050", Justification = "Leave me alone")]
+    [RequiresUnreferencedCode("Distributed transactions are not compatible with trimming. Correctness of the application cannot be guaranteed after trimming.")]
```
```suggestion
    [RequiresUnreferencedCode("Distributed transactions support is not compatible with trimming. Correctness of the application cannot be guaranteed after trimming.")]
```

eerhardt committed:
Update src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs

roji reviewed (approved):
OK from my side on any strategy picked here.

roji commented on src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:
```diff
@@ -47,7 +47,30 @@ internal DtcProxyShimFactory(EventWaitHandle notificationEventHandle)
 
-    [UnconditionalSuppressMessage("Trimming", "IL2050", Justification = "Leave me alone")]
+    [RequiresUnreferencedCode("Distributed transactions support is not compatible with trimming. Correctness of the application cannot be guaranteed after trimming.")]
+    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(ITransactionDispenser))]
```
Is it worth leaving all these in, given that they're insufficient to make trimming work?

eerhardt replied:
My thinking was that if someone was actually trying to use this functionality in a trimmed app, and they were trying to make it work, they wouldn't need to do this manually themselves. Yes, it isn't guaranteed to work, but we can at least reduce the steps someone would need to do to make it work if they really wanted to.

roji replied:
The thing is that unless I'm mistaken, a trimmed app won't work because of the COM interop, and I'm not sure it's possible for a user to make it work (because of https://github.com/dotnet/runtime/pull/75176#discussion_r964340209) (so it's not about guaranteeing or not - I think we know it won't work). If that's indeed the case, then these may provide some false impression that things work (or can be made to work) when that's not the case.

But am OK merging these in any case.

roji replied:
Something like the following should trigger this code path:

```c#
using (var scope = new TransactionScope())
using (var conn1 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
using (var conn2 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
{
    conn1.Open();
    conn2.Open();
    
    Console.WriteLine("Escalated to distributed transaction. Press enter to commit.");
    Console.ReadLine();
    
    scope.Complete();
}
```

Note that it's definitely a minimal app that doesn't exercise all the distributed transaction code paths, but if this works it's definitely a start.

eerhardt replied:
Thanks! This helped me test the code as well and caught that I wasn't using all the right DynamicDependencies. I've added a couple more to the interfaces I missed. With that change, I can now publish and successfully run this app:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishTrimmed>true</PublishTrimmed>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Data.SqlClient" Version="5.0.0" />
  </ItemGroup>

</Project>
```

```C#
using Microsoft.Data.SqlClient;
using System.Transactions;

using (var scope = new TransactionScope())
using (var conn1 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
using (var conn2 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
{
    conn1.Open();
    conn2.Open();

    Console.WriteLine("Escalated to distributed transaction. Press enter to commit.");
    Console.ReadLine();
    Console.WriteLine("Completing.");

    scope.Complete();
}

Console.WriteLine("Done.");
```

The only thing someone would need to do is set `<BuiltInComInteropSupport>true</BuiltInComInteropSupport>` on top of `PublishTrimmed=true`.

roji replied:
Great! Do we think this "guarantees" that it works? :)

I think the following interfaces are still missing, for other non-basic flows ([the list of COM interfaces is here](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces)):

* IPrepareInfo
* ITransactionCloner
* ITransactionPhase0EnlistmentAsync
* ITransactionPhase0Factory
* ITransactionPhase0NotifyAsync

Assuming we still want the RequiresUnreferencedCode just in case, we may want to change the message to something like "Distributed transactions support may not be compatible with trimming. If your program creates a distributed transaction via System.Transactions, the correctness of the application cannot be guaranteed after trimming.")]

Thanks again for spending time on this!

/cc @AaronRobinsonMSFT

eerhardt replied:
> Great! Do we think this "guarantees" that it works? :)

No. There isn't a guarantee here until someone rewrites all of this to use COM Wrappers. That's why the warning is in place. I also didn't test it with NativeAOT, which I don't believe would work (probably at all, even with the BuiltInComInteropSupport=true).

> I think the following interfaces are still missing

I can add those and sort the list.

> we may want to change the message to something like

I will update the message.

roji replied:
Thanks, sounds good!

MichalStrehovsky replied:
I'll leave this up to Aaron because any bugs with this will inevitably land on his plate, but I'm a bit uneasy adding extra annotations so that the very basic scenario works and leaving everything else untested because there's a warning and things are allowed not to work.

I don't like when we generate useless warnings because it desensitizes users to them.

I would lean towards either having a warning and doing _nothing_ to make things works with trimming, or ensuring this works and declare it a supported scenario with no warnings. I would lean towards the former.

eerhardt replied:
> ensuring this works and declare it a supported scenario with no warnings

Since the only way to do that is to convert all this to COMWrappers, this would take a lot more work than worth it at this point. So I'd say that option is off the table in the near term.

> I don't like when we generate useless warnings because it desensitizes users to them.

The scenario still doesn't work by default with trimming. The user still needs to manually enable BuiltInComInteropSupport before it will work. So it isn't useless.

> having a warning and doing nothing to make things works with trimming

In the end, if a customer is actually trying to make this work, I think these annotations help them. They don't need to find this list themselves and somehow figure out how to reference all these internal types in either their own DynamicDependency attributes or in a ILLink.Descriptors.xml file.

MichalStrehovsky replied:
> The scenario still doesn't work by default with trimming. The user still needs to manually enable BuiltInComInteropSupport before it will work. So it isn't useless.

The message in the RUC attribute doesn't mention COM.

> In the end, if a customer is actually trying to make this work, I think these annotations help them.

They help them on the golden path that was tested. I don't know how wide that path is.

In general it feels better when unsupported things crash the moment `PublishTrimmed` is enabled on a project (or S.T.Local references are added) than if they appear to work and hit the _very obscure_ COM failure modes long after `PublishTrimmed` was enabled on the project and people already forgot about it and got desensitized to the warning. If we don't try to help them, most people will give up on trimming with this assembly and that honestly feels like a better outcome.

eerhardt replied:
> The message in the RUC attribute doesn't mention COM.

That was intentional. I left it vague because like you said there may be other obscure holes waiting around the corner. I didn't want to mention "to make it work, enable COM".

> most people will give up on trimming with this assembly and that honestly feels like a better outcome.

I think this the root of where we disagree. If someone needs to trim their app, I'd rather get them part of the way there, if possible/reasonable. There's a reason they are trying to do it. It doesn't seem like a good customer experience to not do these simple things to help them.

MichalStrehovsky replied:
> I think this the root of where we disagree. If someone needs to trim their app, I'd rather get them part of the way there, if possible/reasonable

I usually have the same position, but draw a very strong line for COM. The failure modes are undiagnosable for a pure .NET developer.

eerhardt replied:
Since I really want the other changes in this PR in, I've removed the DynamicDependency attributes. They are lower importance than:

1. Getting the developer a warning when using this code in trimmed/NativeAOT apps
2. Removing the size regression on Windows when using `partial` trimming

eerhardt committed:
Add more DynamicDependencies to make trimming work

eerhardt committed:
PR feedback

- Complete the list of all transaction interfaces
- Sort the DynamicDependencies
- Update the trim warning message

eerhardt committed:
Remove the DynamicDependencies attributes

This gives a better message to users that distributed transactions don't work with trimming.

eerhardt commented:
@vitek-karas @jkurdek - how do I fix the trimming error?

```
  sfx -> Trimming win-x86 shared framework assemblies with ILLinker...
D:\a\_work\1\s\artifacts\bin\ILLinkTrimAssembly\net7.0-windows-Release-x86\suppressions-xmls\ILLink.Suppressions.LibraryBuild.System.Transactions.Local.xml(4,6): Trim analysis error IL2121: Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2026'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
##[error]artifacts\bin\ILLinkTrimAssembly\net7.0-windows-Release-x86\suppressions-xmls\ILLink.Suppressions.LibraryBuild.System.Transactions.Local.xml(4,6): error IL2121: (NETCORE_ENGINEERING_TELEMETRY=Build) Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2026'. Consider removing the unused warning suppression.
D:\a\_work\1\s\artifacts\bin\ILLinkTrimAssembly\net7.0-windows-Release-x86\suppressions-xmls\ILLink.Suppressions.LibraryBuild.System.Transactions.Local.xml(4,6): Trim analysis error IL2121: Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2026'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
##[error]artifacts\bin\ILLinkTrimAssembly\net7.0-windows-Release-x86\suppressions-xmls\ILLink.Suppressions.LibraryBuild.System.Transactions.Local.xml(4,6): error IL2121: (NETCORE_ENGINEERING_TELEMETRY=Build) Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2026'. Consider removing the unused warning suppression.
```

Don't I need a `ILLink.Suppressions.LibraryBuild.xml` file? It looks like it is only happening on x86 builds, for some reason.

vitek-karas commented:
It's because of this:
```C#
        if (RuntimeInformation.ProcessArchitecture == Architecture.X86)
        {
            throw new PlatformNotSupportedException(SR.DistributedNotSupportOn32Bits);
        }
```

The constant prop/branch removal will effectively remove the rest of the method on X86, and thus the suppression is considered redundant because the code it relates to isn't there anymore.

This is a known limitation of the redundant suppression detection - it doesn't work correctly with branch removal (it sees the code after branch removal). 

There were 2 or 3 other places where we hit this in the runtime. The workaround is to extract the rest of the method after the if into a separate method and point the suppression to that. On X86 that new method will be completely removed and there won't be any warnings (the redundant suppression detection doesn't trigger on unmarked methods at all).

eerhardt committed:
Fix x86 build. Move the ILLink suppression to a method that is completely trimmed on x86.

eerhardt commented:
Test failure is https://github.com/dotnet/runtime/issues/490.
